### DISCUSSION TITLE
Convert the simulated flow field data into vtk format files.

### DISCUSSION THREAD

**Discussion by Zhiming-Cai** [NatLabRockies/floris#889] [Q&A] [answered]
I plan to use Floris-v3.5 to convert the simulated flow field data into vtk format files. What should I do?

**Comment by rafmudaf** [comment]
Hi @CaiZhiming-HHU thanks for reaching out. It's not clear what specifically you're asking. What have you tried, and what is your plan for making this conversion?

**Reply by Zhiming-Cai** [reply]
I'm sorry I didn't make myself clear.
I am using floris-v3.5 to simulate a wind farm. I want to convert my simulated flow field data (including wind turbine wake) generated by floris-v3.5 into a vtk file. I think I should use the ‘save_as_vtk’ function in the ‘floris/tools/flow_data.py’ file, but I don't understand how to use this function.
Can you show an example of converting flowfield data generated by floris simulation to vtk format?
Thank you so much!

**Reply by rafmudaf** [reply]
Ah, I see, thanks for the added information. So, you'll first need to create an instance of the `FlowData` class and then use it to call the `save_as_vtk` function. I think this has not been used in a while so it might be out of date - @paulf81 do you know whether this should work for v3.5?

In any case, you could try something like this:
```python
fi = FlorisInterface(...)
fi.calculate_wake()

flow_data = FlowData(
  x=fi.floris.grid.x_sorted[0,0],
  y=fi.floris.grid.y_sorted[0,0],
  z=fi.floris.grid.z_sorted[0,0],
  u=fi.floris.flow_field.u_sorted[0,0],
  v=fi.floris.flow_field.v_sorted[0,0],
  w=fi.floris.flow_field.w_sorted[0,0]
)
flow_data.save_as_vtk()
```

If you try this out, please let me know how it works.

**Reply by Zhiming-Cai** [reply]
Thank you for your prompt reply!
Following the example you gave me, I encountered the following error:
![image](https://github.com/NREL/floris/assets/89171153/b7dcf0c5-e244-4f27-9d42-b30bd4d3ba40)

**Reply by rafmudaf** [reply]
Ah ok, then that answers the question whether this routine is up to date with v3.5. For what it's worth, it has been subsequently removed since v3.6. Rather than debugging and updated v3.5, I suggest to migrate to v4 and consider proposing a similar method for v4. I'm happy to coordinate with you in the development of that.

**Reply by Zhiming-Cai** [reply]
That is to say, the 'save_as_vtk' function was removed in versions after v3.5, right?
This means that I now have no way to convert the wind field file generated by the floris simulation to vtk format?
Thanks again for your reply! If there is a chance, I am very willing to cooperate with you!

**Reply by rafmudaf** [reply, marked as answer]
The entire "flow_data.py" file was removed in https://github.com/NREL/floris/pull/736 in an effort to clean up outdated and unused modules, so you're correct this is was removed after v3.5. Within FLORIS, there is no method to convert to VTK format.

That being said, once you understand the VTK format that you want to use, getting the data from FLORIS is straightforward. I'm happy to work with you on this. If you outline what data you need, I'll answer any questions around how to get it out of FLORIS. And like I mentioned earlier, I suggest to use v4 for this since v3 will be minimally maintained going forward.

**Reply by Zhiming-Cai** [reply]
Ok, I understand. Thank you!
In fact, I am also a user of v4. As you said, I hope to generate vtk file of the whole flow field, but there is no corresponding function found in v4. I can write a csv format file that stores scatter coordinates, and then use some functions used to extract 3D wind speed data in v4 to iterate and extract wind speed at all scatter points. But I think this is too clumsy, and it takes a lot of computing resources. 
So I found v3.5 with the 'save_as_vtk' function, but after a series of attempts, it still didn't achieve my expectations. Since there are so many aspects to try, I will not show you, including the reference to #187. 
Finally, I am glad that you can reply me so efficiently, and I will continue to study and continue to use v4 according to your suggestion. By the way, FLORIS is iterating fast, you guys are great!

**Reply by rafmudaf** [reply]
Ah yes, #187 is from an effort a few years ago to make a web-based dashboard with Plotly, and visualizing the 3d flow field data was included in it somewhat successfully.
Thanks for the words of support, and stay in touch if you want to work together on this.
